packages/core/src/interpreter.ts

```diff
@@ -129,160 +107,132 @@ export class Interpreter<
-  public name: string;
   public ref: ActorRef<TEvent>;
+  private _actorContext: ActorContext<TEvent, SnapshotFrom<TBehavior>>;
 
   /**
    * The globally unique process ID for this invocation.
    */
   public sessionId: string;
-  private forwardTo: Set<string> = new Set();
+
+  // TODO: remove
+  public _forwardTo: Set<AnyActorRef> = new Set();
 
   /**
-   * Creates a new Interpreter instance (i.e., service) for the given machine with the provided options, if any.
+   * Creates a new Interpreter instance (i.e., service) for the given behavior with the provided options, if any.
    *
-   * @param machine The machine to be interpreted
+   * @param behavior The behavior to be interpreted
    * @param options Interpreter options
    */
-  constructor(
-    public machine: StateMachine<
-      TContext,
-      TEvent,
-      any,
-      any,
-      TResolvedTypesMeta
-    >,
-    options?: InterpreterOptions
-  ) {
+  constructor(public behavior: TBehavior, options?: InterpreterOptions) {
     const resolvedOptions = {
       ...defaultOptions,
       ...options
-    } as Required<InterpreterOptions>;
+    };
 
     const { clock, logger, parent, id } = resolvedOptions;
+    const self = this;
 
-    const resolvedId = id !== undefined ? id : machine.key;
-
-    this.name = this.id = resolvedId;
+    // TODO: this should come from a "system"
+    this.sessionId = registry.bookId();
+    this.id = id ?? this.sessionId;
     this.logger = logger;
     this.clock = clock;
     this._parent = parent;
-
     this.options = resolvedOptions;
-
     this.ref = this;
-
-    this.sessionId = this.ref.name;
-  }
-
-  public get initialized() {
-    return this.status === InterpreterStatus.Running;
-  }
-
-  public get initialState(): State<TContext, TEvent, TResolvedTypesMeta> {
-    const initialState =
-      this._initialState ||
-      ((this._initialState = this.machine.getInitialState()),
-      this._initialState);
-    return initialState;
+    this._actorContext = {
+      self,
+      id: this.id,
+      sessionId: this.sessionId,
+      logger: this.logger,
+      exec: (fn) => {
+        if (self.status === InterpreterStatus.NotStarted) {
+          this._deferred.push(fn);
+        } else {
+          fn();
+        }
+      },
+      defer: (fn) => {
+        this._deferred.push(fn);
+      },
+      observers: this.observers
+    };
   }
 
-  public get state(): State<TContext, TEvent, TResolvedTypesMeta> {
-    return this._state!;
-  }
+  // array of functions to defer
+  private _deferred: Array<(state: any) => void> = [];
 
-  /**
-   * Executes the actions of the given state, with that state's `context` and `event`.
-   *
-   * @param state The state whose actions will be executed
-   */
-  public execute(state: State<TContext, TEvent, TResolvedTypesMeta>): void {
-    for (const action of state.actions) {
-      this.exec(action, state);
-    }
+  public getInitialState(): InternalStateFrom<TBehavior> {
```

Comment:
I don't think this is needed on the interpreter - at least not as a public method. The consumer should only care about `.getSnapshot()`

Reply:
This replaces `.initialState` which would have introduced more problems. Should we remove `.getInitialState()` altogether since it wouldn't actually reflect the initial state of a machine started at an arbitrary state?

This has implications for e.g. `@xstate/react`, where we need to know the initial state before starting the machine.

Reply:
> This has implications for e.g. @xstate/react, where we need to know the initial state before starting the machine.

Right, but that's already available through `.getSnapshot()`:
https://github.com/statelyai/xstate/blob/d26b760dbae052d33d2b7366e6b934d608e729f5/packages/core/src/interpreter.ts#L470-L477

I don't think there is a need to expose the initial state separately on the interpreter itself.